AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m

```diff
@@ -0,0 +1,886 @@
+                                                                    */
+                                                                   TVIAudioDeviceReadRenderData(context->deviceContext, audioBuffer, audioBufferSizeInBytes);
+
+                                                               } else {
+
+                                                                   /*
+                                                                    * Return silence when we do not have the playout device context. This is the
+                                                                    * case when the remote participant has not published an audio track yet.
+                                                                    * Since the audio graph and audio engine has been setup, we can still play
+                                                                    * the music file using the AVAudioEngine.
+                                                                    */
+                                                                   memset(audioBuffer, 0, audioBufferSizeInBytes);
+                                                               }
+
+                                                               return bufferList;
+                                                           }];
+        if (!success) {
+            NSLog(@"Failed to set the manual rendering block");
+            return NO;
+        }
+
+        success = [_engine startAndReturnError:&error];
+        if (!success) {
+            NSLog(@"Failed to start AVAudioEngine, error = %@", error);
+            return NO;
+        }
+    }
+    return YES;
+}
+
+- (void)teardownAudioEngine {
+    [self teardownPlayer];
+    [_engine stop];
+    _engine = nil;
+}
+
+- (void)playMusic {
+    if (!_engine) {
+        NSLog(@"Cannot play music. AudioEngine has not been created yet.");
+        return;
+    }
+
+    if (_player.isPlaying) {
+        [_player stop];
+    }
+
+    /*
+     * We will attach an audio player node to AVAudioEnine to the main mixer.
+     * AVAudioPlayerNode -> AVAudioUnitReverb -> MainMixerNode -> Core Audio
+     */
+
+    NSString *fileName = [NSString stringWithFormat:@"%@/%@", [[NSBundle mainBundle] bundlePath], @"mixLoop.caf"];
+    NSURL *url = [NSURL fileURLWithPath:fileName];
+    AVAudioFile *file = [[AVAudioFile alloc] initForReading:url error:nil];
+
+    _player = [[AVAudioPlayerNode alloc] init];
+    _reverb = [[AVAudioUnitReverb alloc] init];
+
+    [_reverb loadFactoryPreset:AVAudioUnitReverbPresetMediumHall];
+    _reverb.wetDryMix = 50;
+
+    [_engine attachNode:_player];
+    [_engine attachNode:_reverb];
+
+    [_engine connect:_player to:_reverb format:file.processingFormat];
+    [_engine connect:_reverb to:_engine.mainMixerNode format:file.processingFormat];
+
+    [_player scheduleFile:file atTime:nil completionHandler:^{}];
```

Comment:
Do we need to teardown the player in the completion handler to detach its node? i.e. What is the effect of tapping "playMusic" after playback completes?

Reply:
we can not teardown player in the completion handler. app crahses.